clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java

```diff
@@ -253,397 +286,114 @@
     private final UnknownServerException unknownServerException = new UnknownServerException("secret");
 
     @Test
-    public void testSerialization() throws Exception {
-        checkRequest(createControlledShutdownRequest(), true);
-        checkResponse(createControlledShutdownResponse(), 1, true);
-        checkErrorResponse(createControlledShutdownRequest(), unknownServerException, true);
-        checkErrorResponse(createControlledShutdownRequest(0), unknownServerException, true);
-        checkRequest(createFetchRequest(4), true);
-        checkResponse(createFetchResponse(true), 4, true);
+    public void testSerialization() {
+        Map<ApiKeys, List<Short>> toSkip = new HashMap<>();
+        // It's not possible to create a MetadataRequest v0 via the builder
+        toSkip.put(METADATA, singletonList((short) 0));
+        // DescribeLogDirsResponse does not have a top level error field
+        toSkip.put(DESCRIBE_LOG_DIRS, DESCRIBE_LOG_DIRS.allVersions());
+        // ElectLeaders v0 does not have a top level error field, when accessing it, it defaults to NONE
+        toSkip.put(ELECT_LEADERS, singletonList((short) 0));
+
+        for (ApiKeys apikey : ApiKeys.values()) {
+            for (short version : apikey.allVersions()) {
+                if (toSkip.containsKey(apikey) && toSkip.get(apikey).contains(version)) continue;
+                AbstractRequest request = getRequest(apikey, version);
+                checkRequest(request);
+                checkErrorResponse(request, unknownServerException);
+                checkResponse(getResponse(apikey, version), version);
+            }
+        }
+    }
+
+    @Test
+    public void testSerializationSpecialCases() {
```

Comment:
Similar to the comment about the class level Javadoc, maybe we should document that these tests are for special cases not covered by `testSerialization()`, so contributors know when they need to add tests to this method?